ocaml/idl/ocaml_backend/gen_server.ml

```diff
@@ -197,33 +196,29 @@ let operation (obj : obj) (x : message) =
   in
   let rbac_check_begin =
     if has_session_arg then
+      let serialize_list lst =
+        String.concat "; " lst |> Printf.sprintf "[%s]"
+      in
+      let serialize_name_list lst =
+        List.map (Printf.sprintf {|"%s"|}) lst |> serialize_list
+      in
+      let default_arg_name_params =
+        if is_non_constructor_with_defaults then
+          List.map (fun dp -> dp.DT.param_name) msg_params_with_default_values
+        else
+          []
+      in
+      let arg_names = orig_string_args @ default_arg_name_params in
+      let key_names = List.map fst x.msg_map_keys_roles in
       [
-        "let arg_names = "
-        ^ List.fold_right
-            (fun arg args -> "\"" ^ arg ^ "\"::" ^ args)
-            orig_string_args
-            ( if is_non_constructor_with_defaults then
-                List.fold_right
-                  (fun dp ss -> "\"" ^ dp.DT.param_name ^ "\"::" ^ ss)
-                  msg_params_with_default_values ""
-                ^ "[]"
-            else
-              "[]"
-            )
-        ^ " in"
-      ; "let key_names = "
-        ^ List.fold_right
-            (fun arg args -> "\"" ^ arg ^ "\"::" ^ args)
-            (List.map (fun (k, _) -> k) x.msg_map_keys_roles)
-            "[]"
-        ^ " in"
+        Printf.sprintf "let arg_names = %s in" (serialize_name_list arg_names)
+      ; Printf.sprintf "let key_names = %s in" (serialize_name_list key_names)
       ; "let rbac __context fn = Rbac.check session_id __call \
          ~args:(arg_names,__params) ~keys:key_names ~__context ~fn in"
       ]
     else
-      ["let rbac __context fn = fn() in"]
+      ["let rbac __context fn = fn () in"]
   in
-  let rbac_check_end = if has_session_arg then [] else [] in
```

Comment:
Would it be possible to use records rather than lists? Or a map? I assume we don't want to have duplicate keys either.

Reply:
Which usage do you mean? for arguments it makes sense to use a map. I don't think it makes sense to use records because each call will have different arguments and the code to process them won't be easily generalizable

Reply:
I was thinking about the list matching and extracting parameters by position. The alternative is to have matches of different lengths and capturing each argument by a named pattern variable.

Reply:
In the end it comes down to serializarion of xmlrpc calls, if arguments is modeled as a list in the xml, the deserialization to a record will still need to extract the default one by position. So this step is unavoidable, whether is done before this match or within it.

Since changing how the calls are modelled in xml means changing the API, it will mean we will have to accept both, which needs serious time and effort, in part because we're not too familiar with this part of xapi.

Reply:
Yes. I think you solved the problem nicely. I also like to use the monadic `let*` for cases like this - you fail, when the structure is unexpected but the regular path is straight. Anyway, fine as is.